Add a function for projecting points onto reference cells.

drwells opened the pull request:
Edit: this PR now returns the projected point instead of computing a distance

This can replace `GeometryInfo::distance_to_unit_cell()` in most circumstances. There is no clear way to define the infinity norm for sloped surfaces (like the top face of a tetrahedron), so this new function uses the Euclidean norm.

I'm motivated by this bit of code in `find_active_cell_around_point()`:

https://github.com/dealii/dealii/blob/81b9727816ecc87f2858e9987e823a9ab02904e6/source/grid/grid_tools.cc#L3144-L3154

this isn't going to work with simplices or mixed meshes and I'd like to get that fixed before the next release.

bangerth commented on include/deal.II/grid/reference_cell.h:
```diff
@@ -588,7 +588,14 @@ class ReferenceCell
+   */
+  template <int dim>
+  double
+  distance_to_unit_cell_square(const Point<dim> &p) const;
```
Maybe `...squared()`? But why not just use the distance?

drwells replied:
I picked this name because we also have `Tensor::norm_square()` and `Point::distance_square()`.

I don't have a good argument either way for why to use the distance or the squared distance. It's a lot easier to minimize the squared distance and since those other two functions exist I figured I would skip the square root. Would you prefer I just implement the not-squared version?

bangerth commented on source/grid/reference_cell.cc:
```diff
@@ -807,6 +807,212 @@ ReferenceCell::gmsh_element_type() const
+namespace
+{
+  double
+  distance_to_line_square(const Point<1> &x0,
+                          const Point<1> &x1,
+                          const Point<1> &p)
+  {
+    return std::min(x0.distance_square(p), x1.distance_square(p));
```
That's not right. If the point is between x0 and x1, the distance should be zero.

drwells replied:
Fixed!

bangerth replied:
It's not fixed. Did you forget to push?

drwells commented:
Your description is absolutely correct and a linear program would be a lot simpler (and I wouldn't have had to spend my morning debugging). However, I see two possible drawbacks with using linear programming here:
1. AFAICT there is no linear programming support in boost and this isn't some trivial algorithm we can implement on our own.
2. The Euclidean norm is more natural than the 1-norm for doing geometric calculations.

bangerth commented:
On second thought: 
* We can shortcut the case where the point is inside, for which there is already an efficient algorithm implemented.
* For the remaining case, I'm pretty sure that the closest point on the boundary of a convex polygon is on a face adjacent to the closest vertex. That cuts down on the number of faces you have to test to 2 (in 2d) and 3 (in 3d). 
* For a face you want to test, compute the vectors that point from the closest vertex to each of the edge-connected vertices. Project the point in question onto the plane spanned by these vectors; if the point of the projection lies in the face, the multiplier for these vectors must be positive; if the point is on the plane but not in the face, one of the multipliers is negative -- in which case the closest vertex is the closest point of the face to the point. I think you can probably implement this in a dimension-independent way if you had a helper function that tells you which other vertices are edge-connected to a given vertex.

drwells commented on source/grid/reference_cell.cc:
```diff
@@ -807,6 +807,256 @@ ReferenceCell::gmsh_element_type() const
+  // Handle simple cases first:
+  Assert(dim != 0, ExcNotImplemented());
+  if (contains_point(p, 0.0))
+    return 0.0;
```
1. Done. We need to do this regardless of how optimized this is since these distance functions don't have a concept of volume (they really just compute a distance to the boundary of the reference cell).

drwells commented on source/grid/reference_cell.cc:
```diff
@@ -807,6 +807,256 @@ ReferenceCell::gmsh_element_type() const
+  if (dim == 2)
+    {
+      for (const unsigned int face_no :
+           faces_for_given_vertex(closest_vertex_no))
```
2. Good idea (see here and the 3D case below)

drwells commented on include/deal.II/grid/reference_cell.h:
```diff
@@ -1200,7 +1207,7 @@ ReferenceCell::n_lines() const
         return 6;
       case ReferenceCells::Pyramid:
-        return 7;
+        return 8;
```
We need some more tests for mixed meshes.

drwells commented on source/grid/reference_cell.cc:
```diff
@@ -807,6 +807,256 @@ ReferenceCell::gmsh_element_type() const
+        }
+
+      // No normal projection was possible, so find the nearest point on a line
+      if (min_distance_square == std::numeric_limits<double>::max())
```
3b. For the sake of completeness: the closest point in 3D may be on a line. In that case we can't do a normal projection.

bangerth commented on source/grid/reference_cell.cc:
```diff
@@ -807,6 +807,256 @@ ReferenceCell::gmsh_element_type() const
+                          const Point<2> &x1,
+                          const Point<2> &p)
+  {
+    const double t =
+      -(x1[0] * x0[0] - std::pow(x0[0], 2) - (x1[0] - x0[0]) * p[0] +
+        x1[1] * x0[1] - std::pow(x0[1], 2) - (x1[1] - x0[1]) * p[1]) /
+      (std::pow(x1[0], 2) - 2 * x1[0] * x0[0] + std::pow(x0[0], 2) +
+       std::pow(x1[1], 2) - 2 * x1[1] * x0[1] + std::pow(x0[1], 2));
```
This is hard to read. Couldn't this be written in the form of some dot products etc? If so, it may also cover the 3d case. I don't see how 2d/3d needs to be treated any different.

Separately, can you add a comment what `t` refers to?

drwells replied:
Fixed!

bangerth reviewed (commented):
I meant to get farther with this, but have to run now :-(

masterleinad reviewed (commented):
Looks OK to me. Changelog entry?

masterleinad commented:
You need to fix
```
/home/runner/work/dealii/dealii/include/deal.II/grid/reference_cell.h:592: warning: expected whitespace after \p command
```

masterleinad commented on include/deal.II/grid/reference_cell.h:
```diff
@@ -588,7 +588,18 @@ class ReferenceCell
-  /*
+  /**
+   * Return the point on the surface of the unit cell closest (in the Euclidean
+   * norm) to @p.
```
```suggestion
   * norm) to @p p.
```

peterrum commented:
```
warning: Inheritance graph for 'AlignedVector' not generated, too many nodes (51), threshold is 50. Consider increasing DOT_GRAPH_MAX_NODES.
/home/runner/work/dealii/dealii/include/deal.II/grid/reference_cell.h:592: warning: expected whitespace after \p command
warning: Inheritance graph for 'SmartPointer' not generated, too many nodes (140), threshold is 50. Consider increasing DOT_GRAPH_MAX_NODES.
warning: Inheritance graph for 'Subscriptor' not generated, too many nodes (190), threshold is 50. Consider increasing DOT_GRAPH_MAX_NODES.
warning: Inheritance graph for 'Table' not generated, too many nodes (50), threshold is 50. Consider increasing DOT_GRAPH_MAX_NODES.
```

bangerth commented on include/deal.II/grid/reference_cell.h:
```diff
@@ -588,7 +588,18 @@ class ReferenceCell
+   */
+  template <int dim>
+  Point<dim>
+  project_to_unit_cell(const Point<dim> &p) const;
```
We've moved away from the term "unit cell" and replaced that by "reference cell". Perhaps rename the function?

drwells replied:
`reference_cell` is redundant here since the class is already named `ReferenceCell` so I went with `closest_point()`.

bangerth commented on tests/grid/project_to_unit_cell.cc:
```diff
@@ -0,0 +1,98 @@
+#include <deal.II/grid/reference_cell.h>
```
This file is missing the standard test header.

bangerth reviewed (commented):
I've got a whole bunch of comments on commentary, but I think the rest is all good.

I realized that we have `ReferenceCell::unit_normal_vectors()` (poorly named, with the plural and all) that, for a given face, provides the normal vector. Would that have been useful in computing the projection?

drwells commented:
Thanks for the feedback. I think I know how to make this function guarantee that the points it computes are in the respective reference cell - I'll try to fix that and the other outstanding issues by the end of the week.

drwells commented on source/grid/reference_cell.cc:
```diff
@@ -807,6 +807,244 @@ ReferenceCell::gmsh_element_type() const
+    const Point<3> shifted_p = p + shift;
+
+    // We are projecting onto a face of a unit cell so the vectors describing
+    // its local coordinate system should be orthogonal. We don't know which of
```
I think your second comment indicates that you saw that this is only for triangles and quadrilaterals - I reworded the comment to hopefully make this clearer.

drwells commented on source/grid/reference_cell.cc:
```diff
@@ -807,6 +807,244 @@ ReferenceCell::gmsh_element_type() const
+                  const Point<3> &               p,
+                  const ReferenceCell            reference_cell)
+  {
+    Assert(reference_cell == ReferenceCells::Triangle ||
+             reference_cell == ReferenceCells::Quadrilateral,
+           ExcNotImplemented());
```
I added another comment to clarify what's going on here to the top-level documentation of the function

drwells commented on source/grid/reference_cell.cc:
```diff
@@ -807,6 +807,244 @@ ReferenceCell::gmsh_element_type() const
+   * with structdim = 2 and spacedim = 3) and the square of that point's
+   * distance to @p.
+   *
+   * If the given point cannot be projected via a normal vector then this
+   * function returns the origin and the largest double precision number.
+   * distance_to_line_square() is in charge of computing the distance to lines.
```
Good question (I added more documentation). This function only projects points along the normal vector of the quad (since that's the shortest distance). However, it might turn out that there is no point on the quad which is on the line parallel to the normal vector which passes through `p`: e.g., for a hexahedron, `p = (2, 2, 2)` cannot be projected to any face via any face normal vector.

drwells commented on source/grid/reference_cell.cc:
```diff
@@ -807,6 +807,244 @@ ReferenceCell::gmsh_element_type() const
+  }
+
+  /**
+   * Compute the nearest point on a quad (in the deal.II sense: i.e., something
+   * with structdim = 2 and spacedim = 3) and the square of that point's
+   * distance to @p.
```
This isn't quite right because it is not a parallelogram when the face reference cell is a `Triangle`. I'll add something equivalent to the documentation.

drwells commented:
> I realized that we have ReferenceCell::unit_normal_vectors() (poorly named, with the plural and all) that, for a given face, provides the normal vector. Would that have been useful in computing the projection?

I don't think so since the algorithm I implemented projects onto the plane defined by the quad and then checks to see if that point is in the quad: the direction happens to be in the normal vector but I never explicitly use it.

drwells commented:
I fixed the outstanding problems and added some extra code to guarantee that the points we find are actually in the reference cell.

When testing the new stuff I found a point for which the current implementation doesn't find the closest point (dealing with the top of the tetrahedron / pyramid is hard) so I need to fix that next. I added it (and some randomly generated points) to the test.

drwells commented:
I fixed a few more problems with tets and pyramids - this is now done!

bangerth commented on include/deal.II/base/std_cxx17/algorithm.h:
```diff
@@ -0,0 +1,37 @@
+
+#include <deal.II/base/config.h>
+
+#ifdef DEAL_II_HAVE_CXX17
+#  include <algorithm>
+#else
+#  include <boost/algorithm/clamp.hpp>
+#endif
+
+DEAL_II_NAMESPACE_OPEN
+namespace std_cxx17
+{
+#ifndef DEAL_II_HAVE_CXX17
+  using boost::algorithm::clamp;
+#else
+  using std::clamp;
+#endif
```
Yes, nice! I've often thought that that must surely be one of the least important additions ever to be added to C++ :-)

bangerth replied:
Oh, absolutely. I'm just surprised the C++ committee spends time on this kind of stuff...

bangerth commented on include/deal.II/base/std_cxx17/algorithm.h:
```diff
@@ -0,0 +1,37 @@
+} // namespace std_cxx17
+DEAL_II_NAMESPACE_CLOSE
+
+#endif // dealii_cxx17_variant_h
```
This is the wrong comment. Please fix this in a follow-up.

bangerth commented on include/deal.II/grid/reference_cell.h:
```diff
@@ -588,7 +588,15 @@ class ReferenceCell
+   */
+  template <int dim>
+  Point<dim>
+  closest_point(const Point<dim> &p) const;
```
Yes, good choice.

bangerth reviewed (approved):
I've got one thing I would like you to fix, but let's do this in a separate PR rather than holding up this PR for even longer. Thanks for indulging my many comments!
